### PR TITLE
IAM-286-Create-GLAuth-Operator

### PR DESCRIPTION
**Pull request by bencekov**
Charmed operator for GLAuth

Implemented:
- Base charm
- config file template
- COS Integration
- Base charm unit test
- Base charm integration test

Missing:
- ldap_public lib charm
- rockcraft.yaml
- Actions
- GLAuth users and groups insertion into the postgresql backend
- Unit tests
- Integration tests

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -187,7 +187,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright 2023 identity-team
+   Copyright 2022 Canonical Ltd.
```
Comment (natalian98) [inline LICENSE:190]: You should add the year in which the repo was first published.
```suggestion
   Copyright 2023 Canonical Ltd.
```

File: README.md [hunk 2]
```diff
@@ -1,26 +1,65 @@
...
+The GLAuth Kubernetes Operator offers integration with the [traefik-k8s-operator](https://github.com/canonical/traefik-k8s-operator). The ingress is provided to the configurable HTTP API interface. Prometheus scraping is available through the HTTP API. 
+
+If you have a traefik deployed and configured in the same model as glauth-k8s, to provide ingress to theAPI run:
+```console
+juju integrate traefik glauth-k8s:ingress
+```
+
+## Canonical Observability Stack
+
+This charm offers integration with observability tools in the [Canonical Observability Stack](https://charmhub.io/topics/canonical-observability-stack).
```
Comment (natalian98) [inline README.md:35]: tbh I would split this pr and have the COS integration added in a separate one.

File: config.yaml [hunk 3]
```diff
@@ -1,14 +1,51 @@
...
+  ldap_port:
+    description: |
+      Port on which LDAP service is accessible
+    type: string
+    default: "3893"
+  http_port:
+    description: |
+      Port on which HTTP services are accessible
+    type: string
+    default: "5555"
```
Comment (natalian98) [inline config.yaml:46]: I'm not sure that we need to make the ports configurable, as we want to maintain only the necessary config options. We don't have that configurable in login ui, for example.

File: metadata.yaml [hunk 4]
```diff
@@ -1,43 +1,30 @@
...
-  httpbin-image:
+  oci-image:
     type: oci-image
-    description: OCI image for httpbin
-    # The upstream-source field is ignored by Juju. It is included here as a reference
-    # so the integration testing suite knows which image to deploy during testing. This field
-    # is also used by the 'canonical/charming-actions' Github action for automated releasing.
-    upstream-source: kennethreitz/httpbin
+    description: GLAuth oci-image
+    upstream-source: glauth_postgres
```
Comment (natalian98) [inline metadata.yaml:16]: There is no such image, which is why the CI is failing

File: src/utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utility functions for the Kratos charm."""
```
Comment (natalian98) [inline src/utils.py:5]:
```suggestion
"""Utility functions for the glauth-k8s charm."""
```

File: config.yaml [hunk 6]
```diff
@@ -1,14 +1,51 @@
...
+  base_dn:
+    description: |
+      Coma separated list of base dns names
     type: string
+    default: "glauth,com"
```
Comment (natalian98) [inline config.yaml:51]:
```suggestion
    default: "glauth.com"
```

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -1,35 +1,100 @@
...
+    app = ops_test.model.applications[APP_NAME]
+
+    await app.scale(3)
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+        wait_for_exact_units=3,
```
Comment (natalian98) [inline tests/integration/test_charm.py:99]: This is not needed because we are waiting for the "full" model to become active anyways.
```suggestion

```

File: tests/unit/test_charm.py [hunk 8]
```diff
@@ -1,68 +1,216 @@
-# Copyright 2023 identity-team
+# Copyright 2022 Canonical Ltd.
```
Comment (natalian98) [inline tests/unit/test_charm.py:1]:
```suggestion
# Copyright 2023 Canonical Ltd.
```

**Review by natalian98 — COMMENTED** [review]
I think the repo should be renamed to glauth-k8s **-operator** (see [this](https://juju.is/docs/sdk/naming#heading--naming-repositories) guideline).

Would it be possible to split this PR and open a separate one for the COS integration? I think it will ease the review process.

File: src/charm.py [hunk 9]
```diff
@@ -1,103 +1,292 @@
...
+            relation_name=self._ingress_relation_name,
+            port=int(self._http_port),
+            strip_prefix=True,
+        )
+
+        self.database = DatabaseRequires(
+            self,
+            relation_name=self._db_relation_name,
+            database_name=self._db_name,
+            extra_user_roles="SUPERUSER",
```
Comment (natalian98) [inline src/charm.py:77]: Could you put the extra user roles in a global variable?

File: src/charm.py [hunk 10]
```diff
@@ -1,103 +1,292 @@
...
+if TYPE_CHECKING:
+    from ops.pebble import LayerDict
 
-# Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
+class GlauthK8SCharm(CharmBase):
+    """Charmed GLAuth."""
```
Comment (natalian98) [inline src/charm.py:41]: maybe
```suggestion
    """Charmed GLAuth Operator for Kubernetes."""
```
?

File: src/charm.py [hunk 11]
```diff
@@ -1,103 +1,292 @@
...
+    def _on_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Event Handler for pebble ready event."""
+        # Necessary directory for log forwarding
+        if not self._container.can_connect():
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting to connect to glauth container")
+            return
+        if not self._container.isdir(str(self._log_dir)):
+            self._container.make_dir(path=str(self._log_dir), make_parents=True)
+            logger.info(f"Created directory {self._log_dir}")
```
Comment (natalian98) [inline src/charm.py:240]: Could you move the comment to this block?
```suggestion
        if not self._container.isdir(str(self._log_dir)):
            # Create a directory for log forwarding
            self._container.make_dir(path=str(self._log_dir), make_parents=True)
            logger.info(f"Created directory {self._log_dir}")
```

File: src/charm.py [hunk 12]
```diff
@@ -1,103 +1,292 @@
...
+    @property
+    def _glauth_service_is_running(self) -> bool:
+        if not self._container.can_connect():
+            return False
+
+        try:
+            service = self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return service.is_running()
```
Comment (natalian98) [inline src/charm.py:130]: this property is not used anywhere

File: src/charm.py [hunk 13]
```diff
@@ -1,103 +1,292 @@
...
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer: LayerDict = {
+            "summary": "GLAuth Application layer",
+            "description": "pebble layer for GLAuth service",
             "services": {
-                "httpbin": {
+                self._container_name: {
```
Comment (nsklikas) [inline src/charm.py:138]: Let's change this variable name to `service_name` (or something along those line) even it has the same value.

File: src/charm.py [hunk 14]
```diff
@@ -1,103 +1,292 @@
...
+            self.unit.status = WaitingStatus("Waiting for glauth service")
+            logger.info("GLAuth service is absent. Deferring database created event.")
+            return
+
+        logger.info("Updating GLAuth config and restarting service")
+        self._container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        self._container.push(self._config_file_path, self._render_conf_file(), make_dirs=True)
+
+        self._container.start(self._container_name)
+        self.unit.status = ActiveStatus()
```
Comment (nsklikas) [inline src/charm.py:273]: Why is this needed?

File: src/utils.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+    This is a hack and should be removed once traefik provides a way for us to
+    request the https URL.
+    """
+    p = urlparse(url)
+
+    p = p._replace(scheme="https")
+    p = p._replace(netloc=p.netloc.rsplit(":", 1)[0])
+
+    return urlunparse(p)
```
Comment (nsklikas) [inline src/utils.py:36]: Why is this needed?

File: src/charm.py [hunk 16]
```diff
@@ -1,103 +1,292 @@
...
+    def _http_port(self) -> str:
+        return self.config["http_port"]
+
+    @property
+    def _basedn(self) -> str:
+        # baseDN example: "dc=glauth,dc=com"
+        dn_input = self.config["base_dn"]
+        list_dns = dn_input.split(",")
+        list_dc = [f"dc={dn}" for dn in list_dns]
+        return ",".join(list_dc)
```
Comment (nsklikas) [inline src/charm.py:165]: Any reason we don't expect the user to define the config with the proper syntax, ie: `dc=example,dc=com`. Since it is a pretty common config for LDAP I don't think we need to be doing this transformation.

File: templates/glauth.cfg.j2 [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  database = "postgres://{{ db_info.get('username') }}:{{ db_info.get('password') }}@{{ db_info.get('endpoints') }}/{{ db_info.get('database_name') }}?sslmode=disable"
+
+[behaviors]
+  IgnoreCapabilities = {{ ignore_capabilities  }}
+  LimitFailedBinds = {{ limited_failed_binds }}
+  NumberOfFailedBinds = {{ number_of_failed_binds }}
+  PeriodOfFailedBinds = {{  period_of_failed_binds }}
+  BlockFailedBindsFor = {{ block_failed_binds_for }}
+  PruneSourceTableEvery = {{ prune_source_tables_every }}
+  PruneSourcesOlderThan = {{ prune_sources_older_than }}
```
Comment (nsklikas) [inline templates/glauth.cfg.j2:24]: Are we sure we want to expose these as config? Isn't it our job as charmers to configure these to make the user's life easier? (cc @massigori)

File: src/charm.py [hunk 18]
```diff
@@ -1,103 +1,292 @@
...
+        self._loki_push_api_relation_name = "logging"
+
+        self._glauth_service_command = f"glauth -c {self._config_file_path}"
+        self._log_dir = Path("/var/log")
+        self._log_path = self._log_dir / "glauth.log"
+
+        self._ingress_relation_name = "ingress"
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [(self.app.name, int(self._ldap_port))]
```
Comment (nsklikas) [inline src/charm.py:64]: Shouldn't we patch the `api` port as well?

**Review by nsklikas — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 18 above)
Comment (bencekov) [reply src/charm.py:64]: So the API port is something we only use for prometheus metrics. I'm under the impression that the http server is something we only want to use if it's behind a proxy. With that being said I can patch that port too.

File: templates/glauth.cfg.j2 (hunk 17 above)
Comment (bencekov) [reply templates/glauth.cfg.j2:24]: On one hand yes, however: there are default values for all these, so users don't have to interact with these configs if they don't want to. If we don't make these configs exportable, then it's really hard to change them for the user.

File: src/charm.py (hunk 16 above)
Comment (bencekov) [reply src/charm.py:165]: I'm somewhat new to LDAP, but I've seen a bunch of docs where people use space instead of a coma, or use all caps for for dc, and ou and such. It would be probably better though to change it, and then normalize the input.

File: src/charm.py [hunk 19]
```diff
@@ -1,103 +1,292 @@
...
+        )
 
     @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+    def _metrics_external_url(self) -> str:
+        return normalise_url(self.ingress.url) if self.ingress.is_ready() else ""
```
[the remark this answers is not in the document]
Comment (bencekov) [reply src/charm.py:119]: The prometheus charm client has the option to use an external url to access scraping features. If there's an ingress relation, then we're passing the ingress url to prometheus, otherwise we pass the empty string instead of None, which is the default value for the external_url parameter.

File: src/charm.py (hunk 14 above)
Comment (bencekov) [reply src/charm.py:273]: when the db is first connected, I use the container's start method. On other db changes and config changes I use the restart method. Not sure what happens when calling restart method on unstarted container. Let me know if you rather I invoked the _handle_status_update_config as response to DataBaseCreatedEvent.

File: src/utils.py (hunk 15 above)
Comment (bencekov) [reply src/utils.py:36]: I assumed this was how we were supposed to format ingress urls based on the hydra and kratos charms.

File: src/charm.py [hunk 20]
```diff
@@ -1,103 +1,283 @@
...
         }
+        return Layer(pebble_layer)
+
+    @property
+    def _ldap_port(self) -> str:
+        return "3893"
+
+    @property
+    def _http_port(self) -> str:
+        return "5555"
```
Comment (natalian98) [inline src/charm.py:148]: Let's store the ports in global variables as we do for other charms.

File: src/charm.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (nsklikas) [reply src/charm.py:121]: Sorry I wasn't clear enough, we are exposing this URL to prometheus, why do we need to "normalise" it?

[the remark this answers is not in the document]
Comment (bencekov) [reply src/charm.py:121]: I assumed normalising the urls was like a new standard of how we share urls. But you're right, it's not necessary here.

File: src/utils.py (hunk 15 above)
Comment (bencekov) [reply src/utils.py:36]: I removed it.

File: src/charm.py (hunk 20 above)
Comment (bencekov) [reply src/charm.py:148]: ok

**Comment by nsklikas** [comment]
Shall we close this in favor of https://github.com/canonical/glauth-k8s-operator/pull/15?
